File: backend/app/api/v1/roles.py
"""Tenant-scoped role management endpoints (FEAT-016, TODO-163).

Base path: /api/v1/tenant/roles
Guards: manage/roles = admin+manager for writes; all staff can read
(list + catalog are open to any authenticated tenant staff).
"""

from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_admin_user, require_permission
from app.db.session import get_session
from app.models.admin_user import AdminUser
from app.schemas.roles import (
    PermissionCatalogItem,
    RoleCreateRequest,
    RolePermissionResponse,
    RoleResponse,
    RoleUpdateRequest,
)
from app.services import roles as role_service

router = APIRouter(prefix="/tenant/roles", tags=["roles"])


def _get_tenant_id(user: AdminUser) -> uuid.UUID:
    if user.tenant_id is None:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="User must belong to a tenant",
        )
    return user.tenant_id


def _parse_uuid(value: str) -> uuid.UUID:
    try:
        return uuid.UUID(value)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found",
        ) from exc


def _to_role_response(role: Any) -> RoleResponse:
    permissions = [
        RolePermissionResponse(action=p.action, resource=p.resource, granted=p.granted)
        for p in sorted(role.permissions, key=lambda p: (p.action, p.resource))
    ]
    return RoleResponse(
        id=role.id,
        name=role.name,
        description=role.description,
        is_system=role.is_system,
        tenant_id=role.tenant_id,
        permissions=permissions,
    )


@router.get("/", response_model=list[RoleResponse])
async def list_roles_endpoint(
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(get_current_admin_user),
) -> list[RoleResponse]:
    """List system roles + this tenant's custom roles. All staff can read."""
    tenant_id = _get_tenant_id(user)
    roles = await role_service.list_roles(session, tenant_id=tenant_id)
    return [_to_role_response(r) for r in roles]


@router.get("/permissions", response_model=list[PermissionCatalogItem])
async def get_permission_catalog_endpoint(
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(get_current_admin_user),
) -> list[PermissionCatalogItem]:
    """Return the full action/resource catalog (label + group). All staff can read."""
    _ = user
    return [PermissionCatalogItem(**item) for item in role_service.get_permission_catalog()]


@router.post(
    "/",
    response_model=RoleResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_role_endpoint(
    body: RoleCreateRequest,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "roles")),
) -> RoleResponse:
    """Create a tenant custom role with its granted permissions."""
    tenant_id = _get_tenant_id(user)
    role = await role_service.create_role(
        session,
        tenant_id=tenant_id,
        name=body.name,
        description=body.description,
        permissions=body.permissions,
        actor_id=user.id,
    )
    return _to_role_response(role)


@router.patch("/{role_id}", response_model=RoleResponse)
async def update_role_endpoint(
    role_id: str,
    body: RoleUpdateRequest,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "roles")),
) -> RoleResponse:
    """Update a role (system permission sets or custom role fields)."""
    tenant_id = _get_tenant_id(user)
    rid = _parse_uuid(role_id)
    role = await role_service.update_role(
        session,
        tenant_id=tenant_id,
        role_id=rid,
        name=body.name,
        description=body.description,
        permissions=body.permissions,
        actor_id=user.id,
    )
    return _to_role_response(role)


@router.delete("/{role_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_role_endpoint(
    role_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "roles")),
) -> Response:
    """Delete a tenant custom role (unassigned only)."""
    tenant_id = _get_tenant_id(user)
    rid = _parse_uuid(role_id)
    await role_service.delete_role(
        session,
        tenant_id=tenant_id,
        role_id=rid,
        actor_id=user.id,
    )
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{role_id}/reset", response_model=RoleResponse)
async def reset_role_defaults_endpoint(
    role_id: str,
    session: AsyncSession = Depends(get_session),
    user: AdminUser = Depends(require_permission("manage", "roles")),
) -> RoleResponse:
    """Reset a system role (manager/employee) to its seeded permission set."""
    tenant_id = _get_tenant_id(user)
    rid = _parse_uuid(role_id)
    role = await role_service.reset_role_defaults(
        session,
        tenant_id=tenant_id,
        role_id=rid,
        actor_id=user.id,
    )
    return _to_role_response(role)

File: frontend/src/lib/api/roles.js
import { apiFetch } from './client.js';

/**
 * Tenant role-management endpoints (staff realm, /api/v1).
 */

/**
 * @typedef {object} RolePermission
 * @property {string} action
 * @property {string} resource
 * @property {boolean} granted
 */

/**
 * @typedef {object} TenantRole
 * @property {string} id
 * @property {string} name
 * @property {string|null} description
 * @property {boolean} is_system
 * @property {string|null} tenant_id
 * @property {RolePermission[]} permissions
 */

/**
 * @typedef {object} PermissionCatalogEntry
 * @property {string} action
 * @property {string} resource
 * @property {string} label
 * @property {string} group
 */

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @returns {Promise<TenantRole[]>}
 */
export function getRoles(fetchFn, token) {
	return apiFetch(fetchFn, '/tenant/roles', { token });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @returns {Promise<PermissionCatalogEntry[]>}
 */
export function getPermissionCatalog(fetchFn, token) {
	return apiFetch(fetchFn, '/tenant/roles/permissions', { token });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @returns {Promise<PermissionCatalogEntry[]>}
 */
export function getProjectPermissionCatalog(fetchFn, token) {
	return apiFetch(fetchFn, '/tenant/roles/project-permissions', { token });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {{ name: string, description?: string|null, permissions: RolePermission[] }} body
 * @returns {Promise<TenantRole>}
 */
export function createRole(fetchFn, token, body) {
	return apiFetch(fetchFn, '/tenant/roles', { method: 'POST', token, body });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} roleId
 * @param {{ name?: string, description?: string|null, permissions?: RolePermission[] }} body
 * @returns {Promise<TenantRole>}
 */
export function updateRole(fetchFn, token, roleId, body) {
	return apiFetch(fetchFn, `/tenant/roles/${roleId}`, { method: 'PATCH', token, body });
}

/**
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} roleId
 * @returns {Promise<null>}
 */
export function deleteRole(fetchFn, token, roleId) {
	return apiFetch(fetchFn, `/tenant/roles/${roleId}`, { method: 'DELETE', token });
}

/**
 * Reset a system role (manager/employee) to its default permissions.
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {string} roleId
 * @returns {Promise<TenantRole>}
 */
export function resetRole(fetchFn, token, roleId) {
	return apiFetch(fetchFn, `/tenant/roles/${roleId}/reset`, { method: 'POST', token });
}
